rpcs3/Emu/Io/Dimensions.cpp

```diff
@@ -0,0 +1,504 @@
+	memcpy(&figure_change_response[6], figure.data.data(), 7);
+	figure_change_response[13] = generate_checksum(figure_change_response, 13);
+	m_figure_added_removed_responses.push(figure_change_response);
+	return true;
+}
+
+bool dimensions_toypad::move_figure(u8 pad, u8 index, u8 old_pad, u8 old_index)
+{
+    // When moving figures between spaces on the portal, remove any figure from the space they are moving to,
+    // then remove them from their current space, then load them to the space they are moving to
+	remove_figure(pad, index, true);
+
+	dimensions_figure& figure = figures[old_index - 1];
+	std::array<u8, 0x2D * 0x04> data = {};
+	memcpy(data.data(), figure.data.data(), figure.data.size());
+	fs::file in_file = std::move(figure.dim_file);
+
+	remove_figure(old_pad, old_index, false);
+
+	load_figure(data, std::move(in_file), pad, index);
+
+	return true;
+}
+
+bool dimensions_toypad::has_figure_been_added_removed() const
+{
+	return !m_figure_added_removed_responses.empty();
+}
+
+std::array<u8, 32> dimensions_toypad::pop_added_removed_response()
+{
+	std::array<u8, 32> response = m_figure_added_removed_responses.front();
+	m_figure_added_removed_responses.pop();
+	return response;
+}
+
+usb_device_dimensions::usb_device_dimensions(const std::array<u8, 7>& location)
+	: usb_device_emulated(location)
+{
+	device = UsbDescriptorNode(USB_DESCRIPTOR_DEVICE, UsbDeviceDescriptor{0x200, 0x0, 0x0, 0x0, 0x20, 0x0E6F, 0x0241, 0x200, 0x1, 0x2, 0x3, 0x1});
+	auto& config0 = device.add_node(UsbDescriptorNode(USB_DESCRIPTOR_CONFIG, UsbDeviceConfiguration{0x29, 0x1, 0x1, 0x0, 0x80, 0xFA}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_INTERFACE, UsbDeviceInterface{0x0, 0x0, 0x2, 0x3, 0x0, 0x0, 0x0}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_HID, UsbDeviceHID{0x0111, 0x00, 0x01, 0x22, 0x001d}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_ENDPOINT, UsbDeviceEndpoint{0x81, 0x03, 0x20, 0x1}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_ENDPOINT, UsbDeviceEndpoint{0x01, 0x03, 0x20, 0x1}));
+}
+
+usb_device_dimensions::~usb_device_dimensions()
+{
+}
+
+void usb_device_dimensions::control_transfer(u8 bmRequestType, u8 bRequest, u16 wValue, u16 wIndex, u16 wLength, u32 buf_size, u8* buf, UsbTransfer* transfer)
+{
+	usb_device_emulated::control_transfer(bmRequestType, bRequest, wValue, wIndex, wLength, buf_size, buf, transfer);
+}
+
+void usb_device_dimensions::interrupt_transfer(u32 buf_size, u8* buf, u32 endpoint, UsbTransfer* transfer)
+{
+	ensure(buf_size == 0x20);
+
+	if (endpoint == 0x81)
+	{
+		// Read Endpoint, it is required to wait until there is a response to reply with
+		// otherwise the game believes that the portal is disconnected. Loop through queries
+		// in a seperate thread, then use fake transfer callback method to push the response
+		// to the list of other fake response
+		std::thread read_message([this, buf, transfer, buf_size]()
```

Comment:
Isn't this adding a potential race condition if two interrupt_transfers happen and the second thread starts before the first?
Also, this unmanaged thread detachment may lead to crashes if the thread is still running when the emulation stops and the objects are destroyed.

Reply:
Only one read message is ever active at a time, so only one of these threads will ever be active at a time. Would there be a better, RPCS3 oriented way to handle this thread?

Reply:
Well. The simplest way may be to make it a member and join it if it's joinable before starting it anew (and in the destructor of the usb_device_dimensions).
You should also set the thread name btw
`thread_base::set_name("The thread name");`

Reply:
Try this : 
<details>
<summary>patch</summary>

```diff
diff --git a/rpcs3/Emu/Cell/lv2/sys_usbd.cpp b/rpcs3/Emu/Cell/lv2/sys_usbd.cpp
index c2b8c8348..fbd1083ad 100644
--- a/rpcs3/Emu/Cell/lv2/sys_usbd.cpp
+++ b/rpcs3/Emu/Cell/lv2/sys_usbd.cpp
@@ -171,20 +171,6 @@ void LIBUSB_CALL callback_transfer(struct libusb_transfer* transfer)
 	usbh.transfer_complete(transfer);
 }
 
-void fake_callback_transfer(UsbTransfer* transfer, u32 buf_size)
-{
-	transfer->fake = true;
-	transfer->expected_count = buf_size;
-	transfer->expected_result = HC_CC_NOERR;
-
-	auto& usbh = g_fxo->get<named_thread<usb_handler_thread>>();
-
-	if (!usbh.is_init)
-		return;
-
-	usbh.push_fake_transfer(transfer);
-}
-
 static void LIBUSB_CALL log_cb(libusb_context* /*ctx*/, enum libusb_log_level level, const char* str)
 {
 	if (!str)
diff --git a/rpcs3/Emu/Io/Dimensions.cpp b/rpcs3/Emu/Io/Dimensions.cpp
index 71af9f231..ca5747466 100644
--- a/rpcs3/Emu/Io/Dimensions.cpp
+++ b/rpcs3/Emu/Io/Dimensions.cpp
@@ -12,8 +12,6 @@
 
 LOG_CHANNEL(dimensions_log, "dimensions");
 
-extern void fake_callback_transfer(UsbTransfer* transfer, u32 buf_size);
-
 dimensions_toypad g_dimensionstoypad;
 
 static constexpr std::array<u8, 16> KEY = {0x55, 0xFE, 0xF6, 0xB0, 0x62, 0xBF, 0x0B, 0x41,
@@ -363,35 +361,32 @@ void usb_device_dimensions::interrupt_transfer(u32 buf_size, u8* buf, u32 endpoi
 
 	if (endpoint == 0x81)
 	{
+		transfer->fake = true;
+		transfer->expected_count = buf_size;
+
 		// Read Endpoint, it is required to wait until there is a response to reply with
 		// otherwise the game believes that the portal is disconnected. Loop through queries
 		// in a seperate thread, then use fake transfer callback method to push the response
 		// to the list of other fake response
-		std::thread read_message([this, buf, transfer, buf_size]()
-			{
-				bool responded = false;
-				do
-				{
-					std::unique_lock lock(query_mutex);
-					std::optional<std::array<u8, 32>> response = g_dimensionstoypad.pop_added_removed_response();
-					if (response)
-					{
-						std::memcpy(buf, response.value().data(), 0x20);
-						responded = true;
-					}
-					else if (!m_queries.empty())
-					{
-						std::memcpy(buf, m_queries.front().data(), 0x20);
-						m_queries.pop();
-						responded = true;
-					}
-					lock.unlock();
-					std::this_thread::sleep_for(100ms);
-				} while (!responded);
-
-				fake_callback_transfer(transfer, buf_size);
-			});
-		read_message.detach();
+		std::unique_lock lock(query_mutex);
+		std::optional<std::array<u8, 32>> response = g_dimensionstoypad.pop_added_removed_response();
+		if (response)
+		{
+			std::memcpy(buf, response.value().data(), 0x20);
+			transfer->expected_result = HC_CC_NOERR;
+		}
+		else if (!m_queries.empty())
+		{
+			std::memcpy(buf, m_queries.front().data(), 0x20);
+			m_queries.pop();
+			transfer->expected_result = HC_CC_NOERR;
+		}
+		else
+		{
+			transfer->expected_count = 0;
+			transfer->expected_result = EHCI_CC_HALTED;
+		}
+		lock.unlock();
 	}
 	else if (endpoint == 0x01)
 	{
```

</details>

Reply:
EHCI_CC_HALTED looks like it reports an error, unsure if this will fix what returning in a separate thread is attempting but will try this out locally

Reply:
I am sorry for doubting you, this worked a treat for me - will add your patch to my branch, do you want author credits for it?

Reply:
No. With this change the top comment is also no longer necessary